packages/react-liveness/src/components/FaceLivenessDetector/service/utils/CustomWebSocketFetchHandler.ts

```diff
@@ -0,0 +1,267 @@
+          socket.readyState as 2 | 3
+        )
+    );
+  }
+
+  private waitForReady(
+    socket: WebSocket,
+    connectionTimeout: number
+  ): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const timeout = setTimeout(() => {
+        this.removeNotUsableSockets(socket.url);
+        reject({
+          $metadata: {
+            httpStatusCode: 500,
+          },
+        });
+      }, connectionTimeout);
+
+      socket.onopen = () => {
+        clearTimeout(timeout);
+        resolve();
+      };
+    });
+  }
+
+  private connect(
+    socket: WebSocket,
+    data: AsyncIterable<Uint8Array>
+  ): AsyncIterable<Uint8Array> {
+    // To notify output stream any error thrown after response
+    // is returned while data keeps streaming.
+    let streamError: Error | undefined = undefined;
+
+    // To notify onclose event that error has occurred.
+    let socketErrorOccurred = false;
+
+    // initialize as no-op.
+    let reject: (err?: unknown) => void = () => {};
+    let resolve: ({
+      done,
+      value,
+    }: {
+      done: boolean;
+      value: Uint8Array;
+    }) => void = () => {};
+
+    socket.onmessage = (event) => {
+      resolve({
+        done: false,
+        value: new Uint8Array(event.data),
+      });
+    };
+
+    socket.onerror = (error) => {
+      socketErrorOccurred = true;
+      socket.close();
+      reject(error);
+    };
+
+    socket.onclose = () => {
+      this.removeNotUsableSockets(socket.url);
+      if (socketErrorOccurred) return;
+
+      if (streamError) {
+        reject(streamError);
+      } else {
+        resolve({
+          done: true,
+          value: undefined as any, // unchecked because done=true.
+        });
+      }
+    };
+
+    const outputStream: AsyncIterable<Uint8Array> = {
+      [Symbol.asyncIterator]: () => ({
+        next: () => {
+          return new Promise((_resolve, _reject) => {
+            resolve = _resolve;
+            reject = _reject;
+          });
+        },
+      }),
+    };
+
+    const send = async (): Promise<void> => {
+      try {
+        for await (const inputChunk of data) {
+          const decodedString = this.utf8decoder.decode(inputChunk);
+          if (decodedString.includes('closeCode')) {
+            const match = decodedString.match(/"closeCode":([0-9]*)/);
+            if (match) {
+              const closeCode = match[1];
+              socket.close(parseInt(closeCode));
+            }
+            continue;
+          }
+
+          socket.send(inputChunk);
+        }
+      } catch (err) {
+        // We don't throw the error here because the send()'s returned
+        // would already be settled by the time sending chunk throws error.
+        // Instead, the notify the output stream to throw if there's
+        // exceptions
+        streamError = err as Error | undefined;
+      } finally {
+        // WS status code: https://tools.ietf.org/html/rfc6455#section-7.4
+        socket.close(1000);
```

Comment:
Is it worth putting this hardcoded value in an enum along with other closure codes we might use?

Reply:
I was just trying to keep this as similar as possible to the original file so I didn't change it. I think with websockets it's well known that 1000 is the success code i guess?